Assignment 3/src/main/java/com/nuclei/assignment/service/graphoperations/GraphOperationsImpl.java

```diff
@@ -0,0 +1,236 @@
+    if (checkNodeExistById(node.getId())) {
+      logger.error(String.format(ExceptionsConstantsUtils.ID_ALREADY_EXISTS,
+          node.getId()));
+      throw new CustomException(String.format(ExceptionsConstantsUtils.ID_ALREADY_EXISTS,
+          node.getId()));
+    }
+    nodes.put(node.getId(), node);
+    System.out.println(SuccessConstantsUtils.CREATED_NODE);
+  }
+  
+  @Override
+  public void createDependency(final int parentId, final int childId) throws CustomException {
+    checkParentAndChildIdPresentInData(parentId, childId);
+    if (checkCyclicDependency(parentId, childId)) {
+      logger.error(String.format(ExceptionsConstantsUtils.CYCLIC_DEPENDENCY,
+          parentId, childId));
+      throw new CustomException(String.format(
+          ExceptionsConstantsUtils.CYCLIC_DEPENDENCY, parentId, childId));
+    }
+  
+    final NodeEntity parentNode = getNodeById(parentId);
+    final NodeEntity childNode = getNodeById(childId);
+  
+    final Set<Integer> parents = childNode.getParents();
+    parents.add(parentId);
+    childNode.setParents(parents);
+  
+    final Set<Integer> children = parentNode.getChildren();
+    children.add(childId);
+    parentNode.setChildren(children);
+    System.out.println(SuccessConstantsUtils.CREATED_DEPENDENCY);
+  
+  }
+  
+  private boolean checkCyclicDependency(final int parentId, final int childId)
+      throws CustomException {
+    final NodeEntity parentNode = getNodeById(parentId);
+    final Queue<NodeEntity> queue = new PriorityQueue<>();
+    final Set<Integer> visitedIds = new HashSet<>();
+    queue.add(parentNode);
+    visitedIds.add(parentId);
+    while (!queue.isEmpty()) {
+      final NodeEntity currentNode = queue.poll();
+      for (final Integer nodeId : currentNode.getParents()) {
+        if (visitedIds.contains(nodeId)) {
+          continue;
+        }
+        if (nodeId.equals(childId)) {
+          return true;
+        }
+        visitedIds.add(nodeId);
+        final NodeEntity tempNode = getNodeById(nodeId);
+        queue.add(tempNode);
+      }
+    }
+    return false;
+  }
+  
+  @Override
+  public Set<Integer> getImmediateParents(final int id) throws CustomException {
+    final NodeEntity node = getNodeById(id);
+    return node.getParents();
+  }
+  
+  @Override
+  public Set<Integer> getImmediateChildren(final int id) throws CustomException {
+    final NodeEntity node = getNodeById(id);
+    return node.getChildren();
+  }
+  
+  @Override
+  public Set<Integer> getAllAncestors(final int id) throws CustomException {
+    final NodeEntity node = getNodeById(id);
+    final Set<Integer> ancestors = new HashSet<>();
+    final Queue<NodeEntity> queue = new PriorityQueue<>();
+    final Map<Integer, Boolean> visited = new ConcurrentHashMap<>();
+    queue.add(node);
+    visited.put(id, true);
+    while (!queue.isEmpty()) {
+      final NodeEntity currentNode = queue.poll();
+      for (final Integer ancestor : currentNode.getParents()) {
```

Comment:
explore functional programming, so that single function can be used for both getAllAncestors and getAllDescendants.